vault/external_tests/raft/raft_binary/raft_test.go

```diff
@@ -47,3 +56,376 @@ func TestRaft_Configuration_Docker(t *testing.T) {
+	if raftWalLeader != "" {
+		oldLeader := leaderNode
+
+		// remove the node
+		leaderNode.Stop()
+
+		// get new leader node
+		leaderNode = cluster.GetActiveClusterNode()
+		leaderClient = leaderNode.APIClient()
+
+		// remove the old leader from the raft configuration
+		removeRaftNode(t, oldLeader, leaderClient, raftWalLeader)
+	}
+
+	// make sure the cluster is healthy
+	stabilize(t, leaderClient)
+
+	// write some data again and read all data from the beginning
+	writeKV(leaderClient, 10)
+	readKV(leaderClient)
+}
+
+// TestRaft_LogStore_Migration_Snapshot checks migration from a boltDB to raftwal
+// by performing a snapshot restore from one cluster to another, and checking no data loss
+func TestRaft_LogStore_Migration_Snapshot(t *testing.T) {
+	t.Parallel()
+	binary := os.Getenv("VAULT_BINARY")
+	if binary == "" {
+		t.Skip("only running docker test when $VAULT_BINARY present")
+	}
+	opts := &docker.DockerClusterOptions{
+		ImageRepo: "hashicorp/vault",
+		// We're replacing the binary anyway, so we're not too particular about
+		// the docker image version tag.
+		ImageTag:    "latest",
+		VaultBinary: binary,
+		ClusterOptions: testcluster.ClusterOptions{
+			NumCores: 1,
+			VaultNodeConfig: &testcluster.VaultNodeConfig{
+				LogLevel: "TRACE",
+			},
+		},
+	}
+	cluster := docker.NewTestDockerCluster(t, opts)
+	defer cluster.Cleanup()
+	rafttest.Raft_Configuration_Test(t, cluster)
+
+	leaderNode := cluster.GetActiveClusterNode()
+	leaderClient := leaderNode.APIClient()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+	defer cancel()
+	err := leaderClient.Sys().MountWithContext(ctx, "kv", &api.MountInput{
+		Type: "kv",
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+	val := 1
+	for i := 0; i < 10; i++ {
+		_, err = leaderClient.Logical().WriteWithContext(ctx, fmt.Sprintf("kv/foo-%d", i), map[string]interface{}{
+			"bar": val,
+		})
+		val++
+	}
+
+	readKV := func(client *api.Client) {
+		for i := 0; i < 10; i++ {
+			secret, err := client.Logical().Read(fmt.Sprintf("kv/foo-%d", i))
+			if err != nil {
+				t.Fatal(err)
+			}
+			if secret == nil || secret.Data == nil {
+				t.Fatal("failed to read the value")
+			}
+		}
+	}
+	readKV(leaderClient)
+
+	// Take a snapshot
+	buf := new(bytes.Buffer)
+	err = leaderClient.Sys().RaftSnapshot(buf)
+	if err != nil {
+		t.Fatal(err)
+	}
+	snap, err := io.ReadAll(buf)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(snap) == 0 {
+		t.Fatal("no snapshot returned")
+	}
+
+	// start a new cluster with raft-wal as its logStore
+	if opts.ClusterOptions.VaultNodeConfig.StorageOptions == nil {
+		opts.ClusterOptions.VaultNodeConfig.StorageOptions = make(map[string]string, 0)
+	}
+	opts.ClusterOptions.VaultNodeConfig.StorageOptions["raft_wal"] = "true"
+
+	// caching the old cluster's barrier keys
+	oldBarrierKeys := cluster.GetBarrierKeys()
+	// clean up the old cluster as there is no further use to it
+	cluster.Cleanup()
```

Comment:
This call has already been deferred from the beginning of the function. This line means `Cleanup()` gets called twice on the original cluster. Is that problematic or no big deal?

Reply:
No it won't cause any issue. It just tries to clean up whatever is left, and if there is nothing left or even had been cleaned up before, it just moves on.